glitch/analysis/security.py

```diff
@@ -112,20 +1254,75 @@ def __check_keyvalue(self, c: CodeElement, name: str,
                             errors.append(Error('sec_def_admin', c, file, repr(c)))
                             break
 
+        def get_au(c, name: str, type: str):
+            if isinstance(c, Project):
+                module_name = os.path.basename(os.path.dirname(file))
+                for m in self.code.modules:
+                    if m.name == module_name:
+                        return get_au(m, name, type)
+            elif isinstance(c, Module):
+                for ub in c.blocks:
+                    au = get_au(ub, name, type)
+                    if au:
+                        return au
+            elif isinstance(c, UnitBlock):
+                for au in c.atomic_units:
+                    if (au.type == type and au.name == name):
+                        return au
+            return None
+
+        def get_module_var(c, name: str):
+            if isinstance(c, Project):
+                module_name = os.path.basename(os.path.dirname(file))
+                for m in self.code.modules:
+                    if m.name == module_name:
+                        return get_module_var(m, name)
+            elif isinstance(c, Module):
+                for ub in c.blocks:
+                    var = get_module_var(ub, name)
+                    if var:
+                        return var
+            elif isinstance(c, UnitBlock):
+                for var in c.variables:
+                    if var.name == name:
+                        return var
+            return None
+
+        # only for terraform
+        var = None
+        if (has_variable and self.tech == Tech.terraform):
+            value = re.sub(r'^\${(.*)}$', r'\1', value)
+            if value.startswith("var."):   # input variable (atomic unit with type variable)
+                au = get_au(self.code, value.strip("var."), "variable")
+                if au != None:
+                    for attribute in au.attributes:
+                        if attribute.name == "default":
+                            var = attribute
+            elif value.startswith("local."):    # local value (variable)
+                var = get_module_var(self.code, value.strip("local."))
```

Comment:
Can you explain what is this doing exactly please? I think you are controlling data flow, i.e., trying to get variables defined in other files, is that correct? If so, this probably should be extracted to other file, so we can create an abstraction in the future somehow. Basically we should have a class with this functionality that receives the technology and for now it only does something when the technology is Terraform.

Reply:
Yes I was controlling data flow. Basically, if the attribute value was stored in a variable it would try to get this value from the variable and check if the value is safe or not.